packages/react-slider/src/components/Slider/useSliderState.tsx

```diff
@@ -1,244 +1,60 @@
+    const newValue = Number(ev.target.value);
+    setCurrentValue(newValue);
 
-  const markPercent = React.useMemo((): string[] => getMarkPercent(markValues), [markValues]);
+    const inputOnChange = state.input.onChange;
+    const propsOnChange = state.onChange;
 
-  const thumbWrapperStyles = {
-    transform: vertical
-      ? `translateY(${valuePercent}%)`
-      : `translateX(${dir === 'rtl' ? -valuePercent : valuePercent}%)`,
-    transition: stepAnimation ? `transform ease-in-out ${animationTime}` : 'none',
-    ...state.thumbWrapper.style,
+    if (inputOnChange && inputOnChange !== propsOnChange) {
+      inputOnChange(ev);
+    } else if (propsOnChange) {
+      propsOnChange(ev, { value: newValue });
+    }
   };
 
-  const trackStyles = {
-    [vertical ? 'top' : dir === 'rtl' ? 'right' : 'left']:
-      origin !== undefined ? `${Math.min(valuePercent, originPercent)}%` : 0,
-    [vertical ? 'height' : 'width']:
+  const rootVariables = {
+    '--slider-thumb-position': valuePercent + '%',
+    '--slider-rail-steps-percent': state.step && state.step > 0 ? `${(step * 100) / (max - min)}%` : '',
+    '--slider-rail-offset': origin !== undefined ? `${Math.min(valuePercent, originPercent)}%` : '0%',
+    '--slider-rail-progress':
       origin !== undefined
         ? `${Math.max(originPercent - valuePercent, valuePercent - originPercent)}%`
         : `${valuePercent}%`,
-    borderRadius: getTrackBorderRadius(),
-    // When a transition is applied with the origin, a visible animation plays when it goes below the min.
-    transition: stepAnimation
-      ? `${vertical ? 'height' : 'width'} ease-in-out ${animationTime}${
-          origin !== undefined
-            ? ', ' + vertical
-              ? 'top'
-              : dir === 'rtl'
-              ? 'right'
-              : 'left' + 'ease-in-out ' + animationTime
-            : ''
-        }`
-      : 'none',
-    ...state.track.style,
   };
 
-  const marksWrapperStyles = marks
-    ? {
-        [vertical ? 'gridTemplateRows' : 'gridTemplateColumns']: markPercent.join(''),
-        ...state.marksWrapper.style,
-      }
-    : {};
-
   // Root props
-  if (!disabled) {
-    state.root.onPointerDown = onPointerDown;
-    state.root.onKeyDown = onKeyDown;
-  }
-
-  // Track Props
-  state.track.style = trackStyles;
-
-  // Mark props
-  if (marks) {
-    state.marksWrapper.children = renderMarks(markValues, marks);
-    state.marksWrapper.style = marksWrapperStyles;
-  }
-
-  // Thumb Wrapper Props
-  state.thumbWrapper.style = thumbWrapperStyles;
-
-  // Active Rail Props
-  state.activeRail.ref = railRef;
+  state.root.style = {
+    ...rootVariables,
+    ...state.root.style,
+  };
 
   // Input Props
-  state.input.ref = useMergedRefs(state.input.ref, inputRef);
   state.input.value = currentValue;
   state.input.min = min;
   state.input.max = max;
   ariaValueText && (state.input['aria-valuetext'] = ariaValueText(currentValue!));
```

Comment:
(I know this isn't new code in this PR, but since this PR is cleaning up the implementation... )

It looks like we only set `aria-valuetext` if the prop is provided. Is that right, or should we always set it, and just stringify the value if no function is provided?

If it's correct as-is, then prefer using `if` when conditionally setting a value. Also, `!` is usually indicative of a bug, or bad types for the variable. It's probably best to guard this with a check, or fix the type for `currentValue`.  
```suggestion
  if (ariaValueText && currentValue !== undefined) {
    state.input['aria-valuetext'] = ariaValueText(currentValue);
  }
```

Reply:
yeah, ariaValueText is something I haven't touched at all from the original implementation.